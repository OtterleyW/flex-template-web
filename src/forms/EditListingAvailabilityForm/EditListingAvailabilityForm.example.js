/* eslint-disable no-console */
import EditListingAvailabilityForm from './EditListingAvailabilityForm';

export const Empty = {
  component: EditListingAvailabilityForm,
  props: {
    onSubmit: values => {
      console.log('Submit EditListingAvailabilityForm with (unformatted) values:', values);
    },
    saveActionMsg: 'Save rules',
    updated: false,
    updateInProgress: false,
  },
  group: 'forms',
};
