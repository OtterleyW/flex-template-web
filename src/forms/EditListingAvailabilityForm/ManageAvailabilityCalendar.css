@import '../../marketplace.css';

:root {
  --ManageAvailabilityCalendar: #d2d2d2;
}

.root {
  @apply --marketplaceH5FontStyles;
  margin-top: 0;
  margin-bottom: 0;
  overflow-x: hidden;
  z-index: 0;

  & :global(.CalendarMonth_caption) {
    @apply --marketplaceH2FontStyles;
    text-align: left;
    padding-bottom: 45px;

    margin-left: 80px;
    margin-top: 0;
    margin-bottom: 0;

    & strong {
      font-weight: 500;
      letter-spacing: 0.2px;
    }
  }

  & :global(.DayPicker) {
    box-shadow: none;
  }

  & :global(.DayPicker__horizontal) {
    background-color: transparent;
    margin-left: -22px;
  }

  & :global(.DayPicker_week-header) {
    top: 55px;

    & small {
      font-size: 13px;
      font-weight: 500;
    }
  }

  & :global(.DayPickerNavigation_leftButton__horizontal),
  & :global(.DayPickerNavigation_rightButton__horizontal) {
    padding: 12px;
    border: solid 1px #e6e6e6;
    background-color: var(--matterColorBright);

    &:hover {

      & svg {
        fill: #000;
      }
    }
    &:focus {
      outline-offset: -4px;
    }
  }

  & :global(.DayPickerNavigation_leftButton__horizontal) {
    border-top-left-radius: 2px;
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
    border-bottom-left-radius: 2px;
  }

  & :global(.DayPickerNavigation_rightButton__horizontal) {
    /* The navigation arrows have 9px padding. Add -9px margin to
    align the arrows with the calendar */
    left: 56px;
    right: unset;
    border-top-left-radius: 0;
    border-top-right-radius: 2px;
    border-bottom-right-radius: 2px;
    border-bottom-left-radius: 0;
  }

  & :global(.CalendarMonthGrid) {
    background-color: transparent;
  }

  & :global(.CalendarMonth) {
    background-color: transparent;
  }

  & :global(.CalendarMonth_table) {
    border: 1px solid var(--ManageAvailabilityCalendar);
  }

  & :global(.CalendarDay__default) {
    border: 0;
    background-color: var(--ManageAvailabilityCalendar);

    &:hover {
      border: 0;
      background-color: var(--ManageAvailabilityCalendar);
    }
  }

  & :global(.CalendarDay__selected) {
    color: var(--matterColor);
  }

  & :global(.DayPickerKeyboardShortcuts_show__bottomRight) {
    right: -20px;
  }
}

@media (--viewportMedium) {
  .root {
    margin-top: 0;
    margin-bottom: 0;

    & :global(.DayPickerNavigation_leftButton__horizontal),
    & :global(.DayPickerNavigation_rightButton__horizontal) {
      background-color: var(--matterColorLight);
    }

  }
}

@media (--viewportLarge) {
  .root {

    & :global(.DayPickerNavigation_leftButton__horizontal),
    & :global(.DayPickerNavigation_rightButton__horizontal) {
      background-color: var(--matterColorLight);
    }
  }
}

.dayWrapper {
  width: 100%;
  height: 100%;
}

.day {
  display: flex;
  width: 100%;
  height: 100%;
  justify-content: flex-end;
  align-items: flex-end;
  padding: 8px;

  &:hover {
  }
}

.dayNumber {
  font-size: 14px;
  line-height: 14px;
}

.default {
  composes: day;
  background-color: white;
}

.past {
  composes: day;
  background-color: #E6E6E6;

  & .dayNumber {
    text-decoration: line-through;
    color: lightgrey;
  }
}

.today {
  composes: day;
  background-color: white;

  & .dayNumber {
    text-decoration: underline;
  }
}

.blocked {
  composes: day;
  background-color: #E6E6E6;
}

.reserved {
  composes: day;
  background-color: #e0f8e7;
}

.rootNextMonthIcon,
.rootPreviousMonthIcon {
  stroke: var(--matterColor);
  fill: var(--matterColor);
}
