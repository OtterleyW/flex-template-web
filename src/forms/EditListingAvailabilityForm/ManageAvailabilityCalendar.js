import React, { Component } from 'react';
import { func, object, string } from 'prop-types';
import { DayPickerSingleDateController, isSameDay, isInclusivelyBeforeDay, isInclusivelyAfterDay } from 'react-dates';
import classNames from 'classnames';
import moment from 'moment';
import { IconArrowHead } from '../../components';

import css from './ManageAvailabilityCalendar.css';

export const HORIZONTAL_ORIENTATION = 'horizontal';

const isPast = day => {
  const today = moment();
  return !isSameDay(day, today) && isInclusivelyBeforeDay(day, today);
};

const isReserved = (reservedDays, day) => {
  return !!reservedDays.find((d) => isSameDay(moment(d), day));
}

const isBlocked = (blockedDays, day) => {
  return !!blockedDays.find((d) => isSameDay(moment(d), day));
}

const handleDayClickWithProps = props => {
  const { availabilityCalendarChanges, currentMonth, onDayAllowed, onDayBlocked } = props;

  const { blockedDaysCurrent = [], blockedDaysAdded = [], reservedDays = [] } = availabilityCalendarChanges[currentMonth] || {};
  const blockedDays = [...blockedDaysCurrent, ...blockedDaysAdded];

  return day => {
    if (isReserved(reservedDays, day) || isPast(day)) {
      // Cannot allow or block a reserved or a past day
      return;
    } else if (isBlocked(blockedDays, day)) {
      onDayAllowed(day);
    } else {
      onDayBlocked(day);
    }
  };
};

class ManageAvailabilityCalendar extends Component {
  constructor(props) {
    super(props);
    this.dayPickerWrapper = null;
    this.dayPicker = null;

    this.state = {
      focused: true,
      date: null,
    };

    this.onDateChange = this.onDateChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
  }

  onDateChange(date) {
    this.setState({ date });

    const handleDayClick = handleDayClickWithProps(this.props);
    handleDayClick(date);
  }

  onFocusChange() {
    // Force the focused states to always be truthy so that date is always selectable
    this.setState({ focused: true });
  }

  render() {
    const {
      className,
      rootClassName,
      currentMonth,
      availabilityCalendar,
      availabilityCalendarChanges,
      onMonthChanged,
      onDayAllowed,
      onDayBlocked,
      ...rest,
    } = this.props;
    const { focused, date } = this.state;

    const { clientWidth: width, clientHeight: height } = this.dayPickerWrapper || { clientWidth: 0, clientHeight: 0 };
    const daySize = width > 744
      ? 100
      : width > 344
        ? Math.floor((width - 44) / 7)
        : 42;

    const { blockedDaysCurrent = [], blockedDaysAdded = [], reservedDays = [] } = availabilityCalendarChanges[currentMonth] || {};
    const blockedDays = [...blockedDaysCurrent, ...blockedDaysAdded];

    const renderDayContents = day => {
      const dayClasses = classNames(css.default, {
        [css.past]: isPast(day),
        [css.today]: isSameDay(moment(), day),
        [css.blocked]: isBlocked(blockedDays, day),
        [css.reserved]: isReserved(reservedDays, day),
      });

      return (
        <div className={css.dayWrapper}>
          <span className={dayClasses}>
            <span className={css.dayNumber}>{day.format('DD')}</span>
          </span>
        </div>
      );
    };

    const currentMontMoment = moment(currentMonth, 'YYYY-MM');
    const classes = classNames(rootClassName || css.root, className);

    return (
      <div className={classes} ref={c => { this.dayPickerWrapper = c; }}>
        {width > 0 ? (
          <DayPickerSingleDateController
            {...rest}
            ref={c => { this.dayPicker = c; }}
            numberOfMonths={1}
            navPrev={<IconArrowHead direction="left" />}
            navNext={<IconArrowHead direction="right" />}
            daySize={daySize}
            initialVisibleMonth={() => moment(currentMonth, 'YYYY-MM')}
            renderDayContents={renderDayContents}
            focused={focused}
            date={date}
            onDateChange={this.onDateChange}
            onFocusChange={this.onFocusChange}
            onPrevMonthClick={() => {
              onMonthChanged(currentMontMoment.subtract(1, 'months').format('YYYY-MM'));
            }}
            onNextMonthClick={() => {
              onMonthChanged(currentMontMoment.add(1, 'months').format('YYYY-MM'));
            }}
            hideKeyboardShortcutsPanel={width < 400}
          />
          ):null}
      </div>
    );
  }
};

ManageAvailabilityCalendar.defaultProps = {
  className: null,
  rootClassName: null,

  // day presentation and interaction related props
  renderCalendarDay: undefined,
  renderDayContents: null,
  isDayBlocked: () => false,
  isOutsideRange: day => !isInclusivelyAfterDay(day, moment()),
  isDayHighlighted: () => false,
  enableOutsideDays: true,

  // calendar presentation and interaction related props
  orientation: HORIZONTAL_ORIENTATION,
  withPortal: false,
  initialVisibleMonth: null,
  numberOfMonths: 2,
  onOutsideClick() {},
  keepOpenOnDateSelect: false,
  renderCalendarInfo: null,
  isRTL: false,

  // navigation related props
  navPrev: null,
  navNext: null,
  onPrevMonthClick() {},
  onNextMonthClick() {},

  // internationalization
  monthFormat: 'MMMM YYYY',
};

ManageAvailabilityCalendar.propTypes = {
  className: string,
  rootClassName: string,
  availabilityCalendar: object.isRequired,
  availabilityCalendarChanges: object.isRequired,
  onDayAllowed: func.isRequired,
  onDayBlocked: func.isRequired,
  onMonthChanged: func.isRequired,
};

export default ManageAvailabilityCalendar;
