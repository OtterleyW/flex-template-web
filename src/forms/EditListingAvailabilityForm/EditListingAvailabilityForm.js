import React, { Component } from 'react';
import { bool, func, object, string } from 'prop-types';
import { compose } from 'redux';
import { Form as FinalForm } from 'react-final-form';
import { intlShape, injectIntl, FormattedMessage } from 'react-intl';
import classNames from 'classnames';
import { propTypes } from '../../util/types';
import { Form, Button } from '../../components';

import ManageAvailabilityCalendar from './ManageAvailabilityCalendar';
import css from './EditListingAvailabilityForm.css';

const today = new Date();
const twoDigitPad = (num) => {
    return num < 10 ? '0' + num : num;
}
const monthString = (today) => `${today.getFullYear()}-${twoDigitPad(today.getMonth() + 1)}`;

const isSameDay = (a, b) => {
  if (!(a instanceof Date && b instanceof Date)) return false;
  // Compare least significant, most likely to change units first
  // Moment's isSame clones moment inputs and is a tad slow
  return a.getDate() === b.getDate()
    && a.getMonth() === b.getMonth()
    && a.getFullYear() === b.getFullYear();
}

export class EditListingAvailabilityFormComponent extends Component {
  constructor(props) {
    super(props);

    this.state = {
      currentMonth: monthString(today),
      availabilityCalendarChanges: {...props.availabilityCalendar},
    };

    this.onDayAllowed = this.onDayAllowed.bind(this);
    this.onDayBlocked = this.onDayBlocked.bind(this);
  }

  onDayAllowed(day) {
    console.log('unblocked', day.format());
    this.setState(prevState => {
      const monthData = prevState.availabilityCalendarChanges[prevState.currentMonth] || [];
      const blockedDaysCurrent = monthData.blockedDaysCurrent || [];
      const newBlockedDays = blockedDaysCurrent.filter(d => {
        return !isSameDay(d, day.toDate());
      });
      return {
        availabilityCalendarChanges: {
          ...prevState.availabilityCalendarChanges,
          [this.state.currentMonth]: {
            ...prevState.availabilityCalendarChanges[this.state.currentMonth],
            blockedDaysCurrent: newBlockedDays,
          },
        },
      };
    });
  }
  onDayBlocked(day) {
    console.log('blocked', day.format());
    this.setState(prevState => {
      const monthData = prevState.availabilityCalendarChanges[prevState.currentMonth] || {};
      const blockedDaysCurrent = monthData.blockedDaysCurrent || [];
      const newBlockedDays = [...blockedDaysCurrent, day.toDate()];
      return {
        availabilityCalendarChanges: {
          ...prevState.availabilityCalendarChanges,
          [this.state.currentMonth]: {
            ...prevState.availabilityCalendarChanges[this.state.currentMonth],
            blockedDaysCurrent: newBlockedDays,
          },
        },
      };
    });
  }

  render(){
    return (
      <FinalForm
        {...this.props}
        render={fieldRenderProps => {
          const {
            className,
            rootClassName,
            disabled,
            handleSubmit,
            onMonthChanged,
            //intl,
            invalid,
            pristine,
            saveActionMsg,
            updated,
            updateError,
            updateInProgress,
            availabilityCalendar,
          } = fieldRenderProps;

          const errorMessage = updateError ? (
            <p className={css.error}>
              <FormattedMessage id="EditListingAvailabilityForm.updateFailed" />
            </p>
          ) : null;

          const classes = classNames(rootClassName || css.root, className);
          const submitReady = updated && pristine;
          const submitInProgress = updateInProgress;
          const submitDisabled = invalid || disabled || submitInProgress;

          return (
            <Form className={classes} onSubmit={handleSubmit}>
              {errorMessage}
              <div className={css.calendarWrapper}>
                <ManageAvailabilityCalendar
                  availabilityCalendar={availabilityCalendar}
                  availabilityCalendarChanges={this.state.availabilityCalendarChanges}
                  currentMonth={this.state.currentMonth}
                  onDayAllowed={this.onDayAllowed}
                  onDayBlocked={this.onDayBlocked}
                  onMonthChanged={month => {
                    console.log(month);
                    this.setState({ currentMonth: month });
                    onMonthChanged(month);
                  }}
                />
              </div>

              <Button
                className={css.submitButton}
                type="submit"
                inProgress={submitInProgress}
                disabled={submitDisabled}
                ready={submitReady}
              >
                {saveActionMsg}
              </Button>
            </Form>
          );
        }}
      />
    );
  }
};

EditListingAvailabilityFormComponent.defaultProps = {
  updateError: null,
  availabilityCalendar: {},
};

EditListingAvailabilityFormComponent.propTypes = {
  intl: intlShape.isRequired,
  onSubmit: func.isRequired,
  saveActionMsg: string.isRequired,
  updated: bool.isRequired,
  updateError: propTypes.error,
  updateInProgress: bool.isRequired,
  availabilityCalendar: object.isRequired,
};

export default compose(injectIntl)(EditListingAvailabilityFormComponent);
