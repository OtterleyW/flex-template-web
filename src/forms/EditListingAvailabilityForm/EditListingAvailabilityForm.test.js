// NOTE: renderdeep doesn't work due to map integration
import React from 'react';
import { renderShallow } from '../../util/test-helpers';
import { fakeIntl } from '../../util/test-data';
import { EditListingAvailabilityFormComponent } from './EditListingAvailabilityForm';

const noop = () => null;

describe('EditListingAvailabilityForm', () => {
  it('matches snapshot', () => {
    const tree = renderShallow(
      <EditListingAvailabilityFormComponent
        intl={fakeIntl}
        dispatch={noop}
        onSubmit={v => v}
        saveActionMsg="Save rules"
        updated={false}
        updateInProgress={false}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
