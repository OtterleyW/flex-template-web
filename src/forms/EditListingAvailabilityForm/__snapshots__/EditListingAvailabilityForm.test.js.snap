// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditListingAvailabilityForm matches snapshot 1`] = `
<ReactFinalForm
  dispatch={[Function]}
  intl={
    Object {
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "now": [Function],
    }
  }
  onSubmit={[Function]}
  render={[Function]}
  saveActionMsg="Save rules"
  updateError={null}
  updateInProgress={false}
  updated={false}
/>
`;
